### PR TITLE
[BUGFIX / FEATURE] [OGR] Allow concurrent edition of Shapefiles and Tabfiles in QGIS & MapInfo

### PR DESCRIPTION
**Pull request by rouault**
- Closes https://hub.qgis.org/issues/14378
- Adds new virtual methods in QgsDataProvider(): enterUpdateMode() and leaveUpdateMode()
  and implement them in the OGR provider. Limited to shapefiles and tabfiles
- Implements QgsOGRProvider:reloadData()
- Robustify OGR provider methods so they don't crash if dataset re-opening fails.

Can we backport this to 2.14 ? Since OSGeo4W now ships with GDAL 2.0, it means that QGIS 2.14 will now open .tab in update mode, preventing MapInfo to open them concurrently, which used to work with GDAL 1.X

### CODE REVIEW

File: src/providers/ogr/qgsogrfeatureiterator.cpp [hunk 1]
```diff
@@ -45,6 +45,10 @@ QgsOgrFeatureIterator::QgsOgrFeatureIterator( QgsOgrFeatureSource* source, bool
   mFeatureFetched = false;
 
   mConn = QgsOgrConnPool::instance()->acquireConnection( mSource->mFilePath );
+  if ( mConn->ds == nullptr )
```
Comment (jef-n) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: ``!mConn->ds`` please.

Comment (rouault) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: OK, I personnaly find == nullptr more explicit, but tastes may vary. Is https://www.qgis.org/en/site/getinvolved/development/qgisdevelopersguide/codingstandards.html the more up-to-date doc to refer to ?

Comment (jef-n) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: I know. ``git grep "== nullptr"`` only show wfs & gml ;)

Comment (rouault) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: (WFS & GML style fixed by  https://github.com/qgis/QGIS/pull/3018/commits/f040a7973eaa6210c41e4391dbd962102484b5e8)

**Comment by rouault** [comment]
Rebased to fix conflict with master

**Comment by NathanW2** [comment]
Massive +1 if others are hapoy

**Comment by rouault** [comment]
Additional precision for completness. The name of the PR is maybe a bit misleading. It should read "concurrent opening". Modification in MapInfo while a dataset is opened by QGIS is not possible. But this was already the case with GDAL 1.X.  But modification in QGIS is possible and MapInfo will see the changes. MapInfo apparently needs an exclusive access to the files to allow saving them. Fixing that isn't in the immediate scope of interest and would require both changes in GDAL and QGIS so that QGIS asks GDAL to close file descriptors when it doesn't need access to the files.

**Comment by rouault** [comment]
Re-rebased on top of latest master. Any comment on this, and about the backportability to 2.14 ?

**Comment by nyalldawson** [comment]
It's not an area of code I'm that familiar with, but the code looks good to me. @wonder-sk @jef-n or @m-kuhn would be better placed to comment.

**Comment by m-kuhn** [comment]
What's the issue with qt5 tests?

**Comment by rouault** [comment]
@m-kuhn Apparently I got a segfault in an older Travis build, but just trying locally, it works and I don't see anything suspicious under Valgrind, so I just re-enabled the test

**Comment by m-kuhn** [comment]
The main reason for disabling tests on Qt5 right now is a problem with NULL in python bindings, so if you check for NULL values somewhere, we'll need to keep it disabled it until it's fixed upstream (pyqt).

**Comment by NathanW2** [comment]
@rouault any issues with this currently or still a work in progress?

**Comment by rouault** [comment]
@NathanW2 Ready to be merged as far as I'm concerned

**Pull request merged**